### PR TITLE
[WIP] Axes 3.0.0. Fixes #215 #223 #233 #236

### LINKED ISSUE #215: Should we use the django signal?
As seen at https://github.com/django/django/commit/7cc4068c4470876c526830778cbdac2fdfd6dc26

### PR DESCRIPTION
**Pull request by camilonova**
Hi everyone.

I just did a bunch of changes to use the Django auth signals, so I suppose a lot of things may have changed.

It is possible to be some broken tests but I already tested on a Django 1.11 project and works pretty fine.

Please feel free to add comments or suggestions, and help me out fixing or improving tests. I won't have much time to work on them in the next couple weeks, but I would like this to be merged soon.

any help is very much appreciated.

### CODE REVIEW

File: axes/attempts.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        return True
+
+    if ip_in_blacklist(ip):
+        return True
+
+    if not is_user_lockable(request):
+        return False
+
+    cache_hash_key = get_cache_key(request)
+    failures_cached = cache.get(cache_hash_key)
```
Comment (cybojenix) [inline axes/attempts.py]: Check LOCK_OUT_AT_FAILURE before all this

Comment (camilonova) [inline axes/attempts.py]: Could you please elaborate more on that?

**Comment by camilonova** [comment]
It seems override_settings is not working on the tests. Can anybody please take a look?

**Comment by cybojenix** [comment]
This is due to everything using axes.settings directly. Basically, all the settings are set at import time, and ignore the fact that django settings are dynamic. 

Take a look at django-appconf. I've never used it before, but it should be a quick fix for you

File: axes/tests.py [hunk 2]
```diff
@@ -111,27 +110,27 @@ def test_failure_limit_many(self):
...
             self.assertContains(response, self.LOGIN_FORM_KEY)
 
         # So, we shouldn't have gotten a lock-out yet.
         # We should get a locked message each time we try again
-        for i in range(0, random.randrange(1, FAILURE_LIMIT)):
+        for i in range(0, random.randrange(1, axes_settings.FAILURE_LIMIT)):
             response = self._login()
             self.assertContains(response, self.LOCKED_MESSAGE, status_code=403)
 
-    @patch('axes.decorators.cache.set', return_value=None)
```
Comment (jorlugaqui) [inline axes/tests.py]: @camilonova why to mock the core cache functions and not the invocation to them in `axes.decorator`? Were the tests hitting the cache somehow?

File: CHANGES.txt [hunk 3]
```diff
@@ -10,6 +10,35 @@ Changes
...
+- Add test for get_cache_key function
+  [jorlugaqui]
+
+- Delete cache key in reset command line
+  [jorlugaqui]
+
+- Add signals for setting/deleting cache keys
+  [jorlugaqui]
+>>>>>>> master
+
```
Comment (jorlugaqui) [inline CHANGES.txt]: @camilonova is this normal? or a merging issue? `>>>>` ???

File: axes/models.py [hunk 4]
```diff
@@ -64,17 +64,25 @@ class AccessAttempt(CommonAccess):
     def failures(self):
         return self.failures_since_start
 
-    def __unicode__(self):
+    def __str__(self):
         return six.u('Attempted Access: %s') % self.attempt_time
 
+    # Only for python 2 support
```
Comment (jorlugaqui) [inline axes/models.py]: Why not to use?

```
from __future__ import unicode_literals
from django.utils.encoding import python_2_unicode_compatible

@python_2_unicode_compatible
class MyClass(object):
    def __str__(self):
        return "Instance of my class"
```

File: axes/models.py [hunk 5]
```diff
@@ -64,17 +64,25 @@ class AccessAttempt(CommonAccess):
...
     )
 
-    def __unicode__(self):
+    def __str__(self):
         return six.u('Access Log for %s @ %s') % (
             self.username, self.attempt_time
         )
+
+    # Only for python 2 support
+    def __unicode__(self):
```
Comment (jorlugaqui) [inline axes/models.py]: Same comment as previous one.

File: axes/signals.py [hunk 6]
```diff
@@ -1,38 +1,162 @@
...
+        AccessLog.objects.filter(
             username=user.get_username(),
             logout_time__isnull=True,
-        ).order_by('-attempt_time')[0:1]
-
-        if access_logs:
-            access_log = access_logs[0]
-            access_log.logout_time = now()
-            access_log.save()
+        ).update(logout_time=timezone.now())
```
Comment (jorlugaqui) [inline axes/signals.py]: I don't know if it makes any difference but: is it possible to localize the time here?

**Comment by jorlugaqui** [comment]
@camilonova Still need to check the tests? I can help :)

File: CHANGES.txt [hunk 7]
```diff
@@ -27,6 +36,7 @@ Changes
 
 - Add signals for setting/deleting cache keys
   [jorlugaqui]
+>>>>>>> master
```
Comment (jwhitlock) [inline CHANGES.txt]: Remove this line.

File: axes/tests.py (hunk 2 above)
Comment (camilonova) [inline axes/tests.py]: Maybe the one who did it (I wasn't) wanted to test the method setting the cache.

Feel free to improve it.

File: axes/models.py (hunk 4 above)
Comment (camilonova) [inline axes/models.py]: Feel free to improve it.

File: axes/models.py (hunk 5 above)
Comment (camilonova) [inline axes/models.py]: Feel free to improve it.

File: axes/signals.py (hunk 6 above)
Comment (camilonova) [inline axes/signals.py]: I doesn't make a difference

**Comment by camilonova** [comment]
@jorlugaqui please do, any help is very much appreciated.

File: axes/tests.py (hunk 2 above)
Comment (jorlugaqui) [inline axes/tests.py]: https://github.com/jazzband/django-axes/pull/241/files

File: axes/models.py (hunk 4 above)
Comment (jorlugaqui) [inline axes/models.py]: https://github.com/jazzband/django-axes/pull/241/files

File: axes/models.py (hunk 5 above)
Comment (jorlugaqui) [inline axes/models.py]: https://github.com/jazzband/django-axes/pull/241/files

**Comment by camilonova** [comment]
@cybojenix It was not that "quick" but it worked. Thanks.

**Comment by camilonova** [comment]
@jazzband/members I need help checking `AXES_COOLOFF_TIME` seems is not working as before.

**Comment by blueyed** [comment]
@camilonova 
Please be more careful about pinging (currently) 278 people, most of them not being involved in this particular project.

**Comment by camilonova** [comment]
@blueyed sure.

**Comment by camilonova** [comment]
@jorlugaqui can you help me checking?

**Comment by jorlugaqui** [comment]
@camilonova, yes, will try to test it this week.

**Comment by jorlugaqui** [comment]
@camilonova tested manually in Django==1.11.4, AXES_COOLOFF_TIME seems to be working, but the tests are complaining because of incorrect status code from admin checking. I'm still investigating it.

**Comment by GitRon** [comment]
Sorry for jangling your nerves.., but any news on this issue? Thx!

**Comment by jorlugaqui** [comment]
@camilonova I tried manually with a different version of Django, this time with 1.11.3 and again seems to be working (using Django app admin).

What I noticed is that probably the test `test_failure_limit_once` is not working properly: This test tries to block the user using admin app, the issue is that after the third chance you don't get blocked, it just gives you the error user/logging message, and then because of that others tests fail in cascade, or at least most of them, since other tests use `test_failure_limit_once`  

If I repeat the process manually in admin app then I got the 403 blocked (expected behavior), so I guess something in test configuration is missed. I tried to figure it out yesterday but I did not succed, I will keep trying.

PWD: Any ideas?

**Comment by aleksihakli** [comment]
@jorlugaqui I think this has been reported as an actual issue somewhere. Looking through the issue history could be fruitful, though I'm not 100% sure I remember correctly :)

**Comment by camilonova** [comment]
Any help in here is **very** appreciated. I do not have much time to work on this right now.

File: axes/apps.py [hunk 8]
```diff
@@ -5,7 +5,7 @@ class AppConfig(apps.AppConfig):
     name = 'axes'
 
     def ready(self):
-        from django.contrib.auth import views as auth_views
+        from django.contrib.auth.views import LoginView
         from axes.decorators import watch_login
 
-        auth_views.login = watch_login(auth_views.login)
+        LoginView.dispatch = watch_login(LoginView.dispatch)
```
Comment (JosephKiranBabu) [inline axes/apps.py]: @camilonova The Debian package for django-axes 2.3.3 has a patch to fix  the tests, which corresponds to #223 

This line can be written as follows using a decorator called `method_decorator` from django.
```
LoginView.dispatch = method_decorator(watch_login)(LoginView.dispatch)
```

See the full patch here.
https://anonscm.debian.org/cgit/freedombox/django-axes.git/commit/?id=c8c8e167e3d1be4754dee04cbbe18ff35e2b57ef

With this change, the tests are passing without the refactoring from `decorated_login` to `decorated_dispatch`. Maybe that breaking api change can be avoided this way.

**Comment by camilonova** [comment]
@JosephKiranBabu Done. Thank you.

**Comment by edicopratik** [comment]
@camilonova when you guys are planning to release new version with django 1.11 support ?

**Comment by camilonova** [comment]
@edicopratik as soon as we make the tests run in this branch. If you want/can help please go ahead.

**Comment by camilonova** [comment]
I think I'm moving on removing the failed tests so we can merge this branch and later on we can improve the test suite.

**Comment by camilonova** [comment]
This will only support python3 and django 1.11 and later

**Pull request merged**